Refer to port using MAC address

cathay4t opened the pull request:
The use case is `create a bond with ports holding specified MAC address
regardless the interface name changed or not after reboots`. This use
case cannot be fulfilled by using nmpolicy which will hard-code the
interface name to network backend configurations.

Enable referring VRF, bond, linux-bridge, OVS bridge port via MAC
address via `match` section.

For bond and VRF, use `ports-config` section.
For linux bridge and OVS bridge, use `ports` section.

Example YAML:

```
interfaces:
- name: bond0
  type: bond
  state: up
  link-aggregation:
    mode: balance-rr
    ports-config:
    - match:
        mac-address: 00:23:45:67:89:1a
        iface-type: ethernet
    - match:
        mac-address: 00:23:45:67:89:1b
        iface-type: ethernet
```

By default, the `identifier` is `name` indicating the port is searched
by interface name. The `mac-address` and `iface-type` will be ignored
when applying with `identifier: name`.

With `identifier: mac-address`, `iface-type` and `mac-address` will be used.
Nmstate will raise error if multiple match found. If `iface-type` is
undefined or set to `unknown`, interface type will be ignored when
searching. The `name` property will be ignored when applying with
`identifier: mac-address`.

The mac address is preferring permanent mac address.

Implementation details:
 * For querying, `include_iface_match_in_controller()` will setup the
   `match` section base on `interface.identifier`.

 * For applying, `MergedInterfaces.resolve_iface_match()` will find the
   interface name base on `match` section and fill the `port.name`
   property for follow up validations.

 * For applying, `sync_iface_match_from_port_to_iface()` will change
   interface identifier base on controller's request on `match` section.

 * For applying, created `NmProfile` holding both `NmConnection` and
   `MergedInterface`, so the reapply action could act on correct
   interface even when the `connection.interface-name` undefined for MAC
   identifier connection.

Unit test cases and integration test cases are included.

Resolves: https://issues.redhat.com/browse/RHEL-1745

liangwen12year commented:
The linked issue is wrong? It is a NetworkManager issue.

cathay4t commented:
Fixed. Thanks.

cathay4t commented:
Please wait a little bit https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/1975

tyll commented:
What's the reason to specify the profile name (I guess this is meant to be the NM profile name?) instead of the MAC address directly? Will this change the semantics of inferfaces[].name to be an NM profile name or interface name?

cathay4t commented:
> What's the reason to specify the profile name (I guess this is meant to be the NM profile name?) instead of the MAC address directly? Will this change the semantics of inferfaces[].name to be an NM profile name or interface name?

Some interface might hold identical MAC address[1], it is hard to find the actual interface using MAC address only. With profiler referring, we can expand that interface matching without changed to bond/bridge/etc section.

The `interface.name` only means profile name when `interface.identifier` is `mac-address`.

[1]: https://learn.microsoft.com/en-us/azure/virtual-network/accelerated-networking-how-it-works#bonding

cathay4t commented:
NM main has patch merged and nmstate test case is passing there. The remaining CI failure is caused by ipsec in NM 1.42, will investigate later.
Please proceed the code review.

liangwen12year commented on rust/src/lib/ifaces/linux_bridge.rs:
```diff
@@ -367,7 +394,15 @@ impl LinuxBridgeConfig {
 pub struct LinuxBridgePortConfig {
     /// The kernel interface name of this bridge port.
+    /// When applying with both `profile-name` and `name` defined, only
+    /// `profile-name` will take effect.
```
Can we add tests to validate the situation that both `profile-name`  and `name` are defined ?

liangwen12year commented on rust/src/lib/ifaces/bond.rs:
```diff
@@ -412,6 +418,33 @@ impl BondInterface {
+            .and_then(|b| b.ports_config.as_deref_mut())
+        {
+            for port_conf in
+                ports_config.iter_mut().filter(|p| p.name.is_empty())
```
Can we raise an error when both `name` and `profile_name` are defined?

cathay4t replied:
we cannot. Nmstate allows applying back the queried state, in queried state we include both profile_name and interface name.

cathay4t replied:
Schema changed. Please check the patch again.

tyll commented:
> > What's the reason to specify the profile name (I guess this is meant to be the NM profile name?) instead of the MAC address directly? Will this change the semantics of inferfaces[].name to be an NM profile name or interface name?
> 
> Some interface might hold identical MAC address[1](https://learn.microsoft.com/en-us/azure/virtual-network/accelerated-networking-how-it-works#bonding), it is hard to find the actual interface using MAC address only. With profiler referring, we can expand that interface matching without changed to bond/bridge/etc section.

The bond port definitions in this example still mention a MAC address. How does this help when multiple interfaces have the same MAC address?

> The `interface.name` only means profile name when `interface.identifier` is `mac-address`.

This does not seem to be within the original sprit of Nmstate. Translating from MAC address to interface seems to be a responsibility for a level on top of Nmstate. Otherwise this is more like configuration than state. Also, it is unclear how this would work with the kernel mode.

tyll reviewed (changes requested):
Please resolve the discussion about whether this is a valid schema for Nmstate first before merging.

cathay4t commented:
> > > What's the reason to specify the profile name (I guess this is meant to be the NM profile name?) instead of the MAC address directly? Will this change the semantics of inferfaces[].name to be an NM profile name or interface name?
> > 
> > 
> > Some interface might hold identical MAC address[1](https://learn.microsoft.com/en-us/azure/virtual-network/accelerated-networking-how-it-works#bonding), it is hard to find the actual interface using MAC address only. With profiler referring, we can expand that interface matching without changed to bond/bridge/etc section.
> 
> The bond port definitions in this example still mention a MAC address. How does this help when multiple interfaces have the same MAC address?

When this use case happens, we can expand our `identifier` to fit the needs, e.g. `identifier: mac+driver`.

My point here is we can expand this matching work in interface level instead of bond/bridge/vrf level, they only need to point to interface defined. How that interface matches with kernel/user space NIC is up to specific interface configurations.



> 
> > The `interface.name` only means profile name when `interface.identifier` is `mac-address`.
> 
> This does not seem to be within the original sprit of Nmstate. Translating from MAC address to interface seems to be a responsibility for a level on top of Nmstate. Otherwise this is more like configuration than state. Also, it is unclear how this would work with the kernel mode.

As long as the property is consistent, it can stay in nmstate. The same debate happens for `copy-ip-from` vs `copy-mac-from`. The `copy-ip-from` cannot be nmstate API because interface IP come and go after attached to bond/bridge/etc, but `copy-mac-from` can be nmstate API as firmware MAC address is consistent. 

The `interface.profile-name` is a consistent property, when referring to a interface through MAC, user don't know the interface name, if we allows `interface.name` to be empty, it would be dramatic changes to our validators.

ffmancera reviewed (changes requested):
I honestly believe this is a better fit for NMPolicy. This is why it was created for, isn't? The same way it was suitable for `copy-ip-from`. If we expose the NM internal details to users like `profile names`, then we the state is a configuration file the same way a keyfile is. Also it makes the state less declarative as "show+apply" won't provide always the same results it will depend on profiles lying around.

Why we cannot use NMPolicy for this use case?

cathay4t commented:
Let's put this controversial on hold till we collect enough use cases and comments to decide the fate of it.

I agree this might violate the declarative design of nmstate and the same feature already supported by nmpolicy.

I have reach out to our potential users for use case collecting.

cathay4t commented:
Please move to https://github.com/nmstate/nmstate/issues/2729 for schema discussion first.

cathay4t commented:
Current patch cannot handle switching from mac-base to name-base. Instead of making that patch bigger, I created https://issues.redhat.com/browse/RHEL-66176 to fix it later.

And https://issues.redhat.com/browse/RHEL-66178 for gc.

cathay4t committed:
Refer to port using MAC address

The use case is `create a bond with ports holding specified MAC address
regardless the interface name changed or not after reboots`. This use
case cannot be fulfilled by using nmpolicy which will hard-code the
interface name to network backend configurations.

Enable referring VRF, bond, linux-bridge, OVS bridge port via MAC
address via `match` section in `ports-config` section.
For linux and OVS bridge, the `ports-config` section is an alias of
pre-exist `port`.

Example YAML:

```
interfaces:
- name: bond0
  type: bond
  state: up
  link-aggregation:
    mode: balance-rr
    ports-config:
    - match:
        mac-address: 00:23:45:67:89:1a
        iface-type: ethernet
    - match:
        mac-address: 00:23:45:67:89:1b
        iface-type: ethernet
```

By default, the `identifier` is `name` indicating the port is searched
by interface name. The `mac-address` and `iface-type` will be ignored
when applying with `identifier: name`.

With `identifier: mac-address`, `iface-type` and `mac-address` will be used.
Nmstate will raise error if multiple match found. If `iface-type` is
undefined or set to `unknown`, interface type will be ignored when
searching. The `name` property will be ignored when applying with
`identifier: mac-address`.

The mac address is preferring permanent mac address.

Implementation details:
 * For querying, `include_iface_match_in_controller()` will setup the
   `match` section base on `interface.identifier`.

 * For applying, `MergedInterfaces.resolve_iface_match()` will find the
   interface name base on `match` section and fill the `port.name`
   property for follow up validations.

 * For applying, `sync_iface_match_from_port_to_iface()` will change
   interface identifier base on controller's request on `match` section.

 * For applying, created `NmProfile` holding both `NmConnection` and
   `MergedInterface`, so the reapply action could act on correct
   interface even when the `connection.interface-name` undefined for MAC
   identifier connection.

Unit test cases and integration test cases are included.

Resolves: https://issues.redhat.com/browse/RHEL-1745

Signed-off-by: Gris Ge <fge@redhat.com>

liangwen12year commented on doc/architecture_decisions/ADR-000-adr-template.md:
```diff
@@ -0,0 +1,34 @@
+### Worse
+
+What's the worse outcome could this design brings, please also explain
+why it is not veto this ADR.
```
Do you mean "please explain why this ADR is not vetoed", honestly, it is better to rephrase it a bit for better clarity.

cathay4t commented:
Closing in the favoer of https://github.com/nmstate/nmstate/pull/2829
